modepy/matrices.py

```diff
@@ -265,15 +276,64 @@ def modal_mass_matrix_for_face(face: Face, face_quad: Quadrature,
 
 def nodal_mass_matrix_for_face(face: Face, face_quad: Quadrature,
         trial_functions, test_functions, volume_nodes, face_nodes):
-    """
+    r"""Using the quadrature *face_quad*, provide a matrix :math:`M_f` that
+    satisfies:
+
+    .. math::
+
+        \displaystyle {(M_f)}_{ij} \approx \int_F \phi_i(r) \psi_j(r) dr,
+
+    where :math:`\phi_i` are the (volume) Lagrange basis functions obtained from
+    *test_functions* at *volume_nodes*, and :math:`\psi_j` are the (surface)
+    Lagrange basis functions obtained from *trial_functions* at *face_nodes*.
+
     .. versionadded :: 2020.3
     """
     face_vdm = vandermonde(trial_functions, face_nodes)
     vol_vdm = vandermonde(test_functions, volume_nodes)
 
+    nface_nodes, nface_funcs = face_vdm.shape
+    if nface_nodes != nface_funcs:
+        from warnings import warn
+        warn("nodal_mass_matrix_for_face received a different number of facial "
+                "nodes than facial trial_functions. This used to be accepted, "
+                "but it is not well-defined. This usage will be rejected starting "
+                "in 2022.x.", DeprecationWarning, stacklevel=2)
+
+        face_vdm_inv = la.pinv(face_vdm)
+    else:
+        face_vdm_inv = la.inv(face_vdm)
+
     modal_fmm = modal_mass_matrix_for_face(
             face, face_quad, trial_functions, test_functions)
-    return la.inv(vol_vdm.T).dot(modal_fmm).dot(la.pinv(face_vdm))
+    return la.inv(vol_vdm.T).dot(modal_fmm).dot(face_vdm_inv)
+
+
+def nodal_quad_mass_matrix_for_face(face: Face, face_quad: Quadrature,
+        test_functions, volume_nodes):
+    r"""Using the quadrature *face_quad*, provide a matrix :math:`M_f` that
+    satisfies:
+
+    .. math::
+
+        \displaystyle (M_f \boldsymbol u)_i = \sum_j w_j \phi_i(r_j) u_j,
+
+    where :math:`\phi_i` are the (volume) Lagrange basis functions obtained from
+    *test_functions* at *volume_nodes*, and :math:`w_i` and :math:`r_i` are the
+    weights and nodes from *face_quad*.
+
+    .. versionadded :: 2021.2
+    """
+    vol_vdm = vandermonde(test_functions, volume_nodes)
+
+    mapped_nodes = face.map_to_volume(face_quad.nodes)
+
+    vol_modal_mass_matrix = np.empty((len(test_functions), len(face_quad.weights)))
```

Comment:
Will this work if `face_quad = mp.Quadrature(np.empty((0, 1)), np.empty((0, 1)))`? I don't think this does. Thinking of the case where the face/face_quadrature is coming from a face group on a point set (dim = 0).

Reply:
> `mp.Quadrature(np.empty((0, 1)), np.empty((0, 1)))`

This is strange to me. It looks like you're passing a 2D array (of shape `(0, 1)` for `weights`. Weights to me are always 1D. Could you explain?

Reply:
Sure. So in one of the diffusion tests in mirgecom, (1D), a `QuadratureSimplexElementGroup` is passed in as a face group (`afgrp`), which has spatial dimension zero, and a quadrature rule coming from here: https://github.com/inducer/meshmode/blob/main/meshmode/discretization/poly_element.py#L275

This is the latest commit I used in the grudge branch: https://github.com/inducer/grudge/pull/62/commits/de29e293e66b920a990c80de7f25165ca396c0ae

I managed to reproduce the error by running this in the mirgecom tests: `py.test test_diffusion.py -k 'test_diffusion_accuracy' -x --pdb`. Seems to be failing only for 1D, so you can reproduce this error using the 1D test parameter: `problem = get_static_trig_var_diff(1)`

Reply:
I have not ruled out my own errors, but I am wondering if we're handling 0-dim facets correctly!

Reply:
(Btw, when discussing errors, please include a traceback.) (You can get a usable one with `--tb=native` in pytest. Pytest's default ones are not usable.) 

I get [this](https://gist.github.com/inducer/2bf979884c444674b4bdd87afe65d2df). Is that what you're seeing?

Reply:
> `mp.Quadrature(np.empty((0, 1)), np.empty((0, 1)))`

I just realized that this is in fact a [code snippet from meshmode](https://github.com/inducer/meshmode/blob/511c8af87b09d4f5a1b39b22498664807b161c49/meshmode/discretization/poly_element.py#L275). I did not know that, and it took me a long time to figure it out.

Reply:
It still doesn't make any sense to me though.

Reply:
> > `mp.Quadrature(np.empty((0, 1)), np.empty((0, 1)))`
> 
> I just realized that this is in fact a [code snippet from meshmode](https://github.com/inducer/meshmode/blob/511c8af87b09d4f5a1b39b22498664807b161c49/meshmode/discretization/poly_element.py#L275). I did not know that, and it took me a long time to figure it out.

Yep, I linked that in my reply above. I managed to track it down to that line, as that was causing the traceback I was seeing.



> (Btw, when discussing errors, please include a traceback.) (You can get a usable one with `--tb=native` in pytest. Pytest's default ones are not usable.)
> 
> ~I get [this](https://gist.github.com/inducer/2bf979884c444674b4bdd87afe65d2df).~ (Nvm. I was on a funny loopy branch, turns out.) I get [this](https://gist.github.com/inducer/a9624228ea8de4582bc81049ae72d2be). Is that what you're seeing?

yes.

Reply:
https://github.com/inducer/meshmode/pull/158

Reply:
> Yep, I linked that in my reply above. I managed to track it down to that line, as that was causing the traceback I was seeing.

OIC. It would have helped if I had seen it. Totally on me. Sorry!

Unrelated, unsolicited Github advice: Before copy-pasting a Github code URL, press `y` (the `y` key). Github's Javascript will swap out the URL for one that contains the actual Github revision. That way links remain stable even when code changes.

Reply:
Can confirm now that all is peaceful and happy with this PR + https://github.com/inducer/meshmode/pull/158